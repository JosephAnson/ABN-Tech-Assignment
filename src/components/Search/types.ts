export interface InputProps {
  id?: string
  name?: string
  value: string
}
