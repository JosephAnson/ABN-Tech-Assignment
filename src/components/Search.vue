<script lang="ts" setup>
interface SelectProps {
  id?: string
  name?: string
  value: string
}

const props = withDefaults(defineProps<SelectProps>(), {
  id: undefined,
  name: undefined,
  value: "",
})

const emit = defineEmits(["update:value", "enter"])

const newValue = useVModel(props, "value", emit, { passive: true })

const emitEnter = (value: string) => emit("enter", value)
</script>

<template>
  <div
    class="flex items-center relative shadow-sm bg-gray-800 sm:text-sm border-gray-300 rounded-md"
  >
    <input
      :id="props.id"
      v-model="newValue"
      type="search"
      :name="props.name"
      class="bg-transparent w-full px-4 block h-8 focus:ring-indigo-500 focus:border-indigo-500"
      placeholder="Search TV shows"
      autocomplete="off"
      @keydown.enter="emitEnter(newValue)"
    />
    <button
      class="flex w-10 h-8 p-1 cursor-pointer z-10 hover:bg-gray-700 rounded-r-md"
      @click="emitEnter(newValue)"
    >
      <span class="i-carbon-search w-full h-full"></span>
      <span class="sr-only">Search Button</span>
    </button>
  </div>
</template>
