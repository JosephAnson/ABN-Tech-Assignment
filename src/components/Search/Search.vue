<script lang="ts" setup>
import { defineEmits, defineProps, withDefaults } from "vue"
import { useVModel } from "@vueuse/core"
import type { InputProps } from "~/components/Search/types"

const props = withDefaults(defineProps<InputProps>(), {
  id: undefined,
  name: undefined,
  value: "",
})

const emit = defineEmits(["update:value", "focus", "blur", "enter"])

const newValue = useVModel(props, "value", emit)
</script>

<template>
  <div
    class="flex items-center relative shadow-sm bg-gray-800 sm:text-sm border-gray-300 rounded-md"
  >
    <input
      :id="props.id"
      :value="newValue"
      type="text"
      :name="props.name"
      class="bg-transparent w-full px-4 block h-8 focus:ring-indigo-500 focus:border-indigo-500"
      placeholder="Search TV shows"
      autocomplete="off"
      @input="$emit('update:value', $event.target.value)"
      @focus="$emit('focus')"
      @blur="$emit('blur')"
      @keydown.enter="$emit('enter', $event.target.value)"
    />
    <span
      class="flex w-10 h-8 p-1 cursor-pointer z-10 hover:bg-gray-700 rounded-r-md"
      @click="$emit('enter', newValue)"
    >
      <span class="i-carbon-search w-full h-full"></span>
      <span class="sr-only">Search Button</span>
    </span>
  </div>
</template>
