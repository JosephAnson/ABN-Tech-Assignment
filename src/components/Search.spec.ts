import userEvent from "@testing-library/user-event"
import { cleanup, render } from "@testing-library/vue"
import { afterEach, describe, expect, test } from "vitest"
import Search from "./Search.vue"

const searchName = "Under the Dome"

describe("Search", () => {
  afterEach(() => cleanup())

  test("should render search", () => {
    const { html } = render(Search, {
      props: {
        value: "",
      },
    })
    expect(html()).toMatchSnapshot()
  })

  test("should render search with all props", () => {
    const { html } = render(Search, {
      props: {
        id: "search-id",
        name: "search-id",
        value: "Under the Dome",
      },
    })
    expect(html()).toMatchSnapshot()
  })

  test("should emit changed value on input enter", async () => {
    const { emitted, getByRole } = render(Search, {
      props: {
        id: "search-id",
        name: "search-id",
        value: "",
      },
    })

    const search = await getByRole("searchbox")
    await userEvent.type(search, `${searchName}`)
    await userEvent.type(search, "{enter}")

    expect((search as HTMLInputElement).value).toEqual(searchName)
    expect(emitted<string>("update:value").at(-1)?.[0]).toEqual(searchName)
    expect(emitted<string>("enter")[0][0]).toEqual(searchName)
  })

  test("should emit changed value on button click", async () => {
    const { emitted, getByRole } = render(Search, {
      props: {
        id: "search-id",
        name: "search-id",
        value: "",
      },
    })

    const search = await getByRole("searchbox")
    const button = await getByRole("button")

    await userEvent.type(search, `${searchName}`)
    await userEvent.click(button)

    expect((search as HTMLInputElement).value).toEqual(searchName)
    expect(emitted<string>("enter")[0][0]).toEqual(searchName)
  })
})
