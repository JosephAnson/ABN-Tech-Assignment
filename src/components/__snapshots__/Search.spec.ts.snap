// Vitest Snapshot v1

exports[`Search > should render search 1`] = `"<div class=\\"flex items-center relative shadow-sm bg-gray-800 sm:text-sm border-gray-300 rounded-md\\"><input type=\\"search\\" class=\\"bg-transparent w-full px-4 block h-8 focus:ring-indigo-500 focus:border-indigo-500\\" placeholder=\\"Search TV shows\\" autocomplete=\\"off\\"><button class=\\"flex w-10 h-8 p-1 cursor-pointer z-10 hover:bg-gray-700 rounded-r-md\\"><span class=\\"i-carbon-search w-full h-full\\"></span><span class=\\"sr-only\\">Search Button</span></button></div>"`;

exports[`Search > should render search with all props 1`] = `"<div class=\\"flex items-center relative shadow-sm bg-gray-800 sm:text-sm border-gray-300 rounded-md\\"><input id=\\"search-id\\" type=\\"search\\" name=\\"search-id\\" class=\\"bg-transparent w-full px-4 block h-8 focus:ring-indigo-500 focus:border-indigo-500\\" placeholder=\\"Search TV shows\\" autocomplete=\\"off\\"><button class=\\"flex w-10 h-8 p-1 cursor-pointer z-10 hover:bg-gray-700 rounded-r-md\\"><span class=\\"i-carbon-search w-full h-full\\"></span><span class=\\"sr-only\\">Search Button</span></button></div>"`;
